src/ast/passes/semantic_analyser.cpp

```diff
@@ -1169,6 +1169,26 @@ void SemanticAnalyser::visit(Call &call)
     return;
   }
 
+  auto CheckArgIsRecordPtr =
+      [](Call &call, size_t index, const std::string &name) {
+        assert(index < call.vargs.size());
+        const auto &type = call.vargs.at(index).type();
+        if (!type.IsPtrTy() || !type.GetPointeeTy() ||
+            !type.GetPointeeTy()->IsRecordTy()) {
+          call.addError() << call.func << "() only supports '" << name
+                          << " *' as the argument (" << type.GetTy()
+                          << " provided)";
+          return false;
+        }
+        if (!type.GetPointeeTy()->IsSameType(CreateRecord(name))) {
+          call.addError() << call.func << "() only supports '" << name
+                          << " *' as the argument ('"
+                          << type.GetPointeeTy()->GetName() << " *' provided)";
+          return false;
+        }
+        return true;
+      };
```

Comment:
Sorry, what I meant was, don't make this whole thing a lambda, just the error logging part and please move it back to inside the `call.func == "socket_cookie"` block e.g.
```suggestion
  auto CheckArgIsRecordPtr =
      [](Call &call, size_t index, const std::string &name) {
        assert(index < call.vargs.size());
        const auto &type = call.vargs.at(index).type();
        auto logError = [&]<typename T>(T provided) {
           call.addError() << call.func << "() only supports '" << name
                          << " *' as the argument (" << provided
                          << " provided)";
        };
        if (!type.IsPtrTy() || !type.GetPointeeTy() ||
            !type.GetPointeeTy()->IsRecordTy()) {
           logError(type.GetTy());
           return;
        }
        if (!type.GetPointeeTy()->IsSameType(CreateRecord(name))) {
          logError(type.GetPointeeTy()->GetName());
          return;
        }
        return;
      };
```